source/common/config/datasource.h

```diff
@@ -25,6 +32,108 @@ std::string read(const envoy::api::v2::core::DataSource& source, bool allow_empt
+ * Callback for async data source.
+ */
+using AsyncDataSourceCb = std::function<void(std::string)>;
+
+class LocalAsyncDataProvider;
+using LocalAsyncDataProviderPtr = std::unique_ptr<LocalAsyncDataProvider>;
+
+class RemoteAsyncDataProvider;
+using RemoteAsyncDataProviderPtr = std::unique_ptr<RemoteAsyncDataProvider>;
+
+class LocalAsyncDataProvider {
+public:
+  LocalAsyncDataProvider(Init::Manager& manager, const envoy::api::v2::core::DataSource& source,
+                         bool allow_empty, Api::Api& api, AsyncDataSourceCb&& callback)
+      : init_target_("LocalAsyncDataProvider", [this, &source, allow_empty, &api, callback]() {
+          callback(DataSource::read(source, allow_empty, api));
+          init_target_.ready();
+        }) {
+    manager.add(init_target_);
+  }
+
+  ~LocalAsyncDataProvider() { init_target_.ready(); }
+
+  /**
+   * Create async data provider for local data source.
+   * @param manager init manager for coordinating initialization
+   * @param source data source configuration
+   * @param allow_empty if the data is allowed empty
+   * @param api "Public" API that different components use to interact with the various system
+   * abstractions
+   * @param callback data provider callback
+   * @return data provider for local data source
+   */
+  static LocalAsyncDataProviderPtr create(Init::Manager& manager,
+                                          const envoy::api::v2::core::DataSource& source,
+                                          bool allow_empty, Api::Api& api,
+                                          AsyncDataSourceCb callback) {
+    return std::make_unique<LocalAsyncDataProvider>(manager, source, allow_empty, api,
+                                                    std::move(callback));
+  }
+
+private:
+  Init::TargetImpl init_target_;
+};
+
+class RemoteAsyncDataProvider : public Config::DataFetcher::RemoteDataFetcherCallback {
+public:
+  RemoteAsyncDataProvider(Upstream::ClusterManager& cm, Init::Manager& manager,
+                          const envoy::api::v2::core::RemoteDataSource& source, bool allow_empty,
+                          AsyncDataSourceCb&& callback)
+      : allow_empty_(allow_empty), callback_(std::move(callback)),
+        fetcher_(std::make_unique<Config::DataFetcher::RemoteDataFetcher>(cm, source.http_uri(),
+                                                                          source.sha256(), *this)),
+        init_target_("RemoteAsyncDataProvider", [this]() { start(); }) {
+    manager.add(init_target_);
+  }
+
+  ~RemoteAsyncDataProvider() { init_target_.ready(); }
+
+  // Config::DataFetcher::RemoteDataFetcherCallback
+  void onSuccess(const std::string& data) {
+    callback_(data);
+    init_target_.ready();
+  }
+
+  // Config::DataFetcher::RemoteDataFetcherCallback
+  void onFailure(Config::DataFetcher::Failure failure) {
+    if (allow_empty_) {
+      callback_(EMPTY_STRING);
+      init_target_.ready();
+    } else {
+      throw EnvoyException(
```

Comment:
You may want to rethink about whether the exception is good or not. RemoteDataFetch doesn't catch it and would omit one statement

Reply:
I moved the statement before the callback. I think throwing exception is consistent with current data source implementation.